src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/TransactionValidator.cs

```diff
@@ -0,0 +1,54 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Health.Fhir.Core.Exceptions;
+using static Hl7.Fhir.Model.Bundle;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public static class TransactionValidator
+    {
+        public static void ValidateTransaction(HashSet<string> resourceIdList, EntryComponent entry)
+        {
+            if (ValidateBundleEntry(entry))
+            {
+                string resourceId = GetResourceUrl(entry);
+
+                if (!string.IsNullOrEmpty(resourceId))
+                {
+                    if (resourceIdList.Contains(resourceId))
+                    {
+                        throw new RequestNotValidException(string.Format(Api.Resources.ResourcesMustBeUnique, entry.Request.Url));
+                    }
+
+                    resourceIdList.Add(resourceId);
+                }
+            }
+        }
+
+        private static bool ValidateBundleEntry(EntryComponent entry)
+        {
+            string requestUrl = entry.Request.Url;
+
+            // Check for duplicate resources within a bundle entry is skipped if the entry is bundle or if the request within a entry is not modifying the resource.
```

Comment:
Can you explain this a little bit. Are you saying that if an entry inside the bundle is another bundle, we won't validate across these two bundles?

Reply:
I think the name might be a little misleading. It's not really validating the bundle entry, but rather it's check to see if validation is needed?

Reply:
> Can you explain this a little bit. Are you saying that if an entry inside the bundle is another bundle, we won't validate across these two bundles?

Yes. For now this is the case.